### DISCUSSION TITLE
Undeprecate session.update method

### DISCUSSION THREAD

**Discussion by mbellade** [hibernate/hibernate-orm#6068] [Design Proposals]
With Hibernate 6 we deprecated the `Session#update` and suggest to use `merge` instead.

While investigating [HHH-16111](https://hibernate.atlassian.net/browse/HHH-16111) I noticed some different behavior regarding `Optimistic Locking` and increase in the `Version` attribute, with particular regards to updates to the entity's relationships.

The user's case consists basically in a parent entity with a `@OneToMany(mappedBy="...")` collection of child entities. The example he provided changes the collection of child entityes on a *detached* parent entity (by either adding/removing or modifying a child) and tries to reattach the state of the parent entity and its relationship. With Hibernate 5, he used to use the `Session#update` methods which triggered an update of the parent entity regardless of changes to the child entities collection, thus increasing the `Version` attribute for optimistic locking.

When upgrading to Hibernate 6, the user switched to using the `Session#merge` - as suggested by the `@deprecated` javadoc - which actually behaves differently in this particular case: since the parent entity is the inverse side of the relationship, and no changes to the parent entity itself were made, `merge` does not increase the `Version` attribute.

It looks like this is in accordance to the JPA spec, which specifies (section [3.4.2. Version Attributes](https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1.html#a2059)):

> The version attribute is updated by the persistence provider runtime when the object is written to the database. All non-relationship fields and properties and all relationships owned by the entity are included in version checks.

And defines the owning-side of a relationship (section [2.10. Entity Relationships](https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1.html#a516)) as:
> The inverse side of a bidirectional relationship must refer to its owning side by use of the mappedBy element of the OneToOne, OneToMany, or ManyToMany annotation. The mappedBy element designates the property or field in the entity that is the owner of the relationship.
> ...

In fact, removing `mappedBy` from the `OneToMany` association (making the parent entity the owning side) does correctly trigger an increase of the parent's entity `Version` even when using the `merge` method.

In conclusion, `update` has a different behavior compared to `merge`: it *forces* an update of the detached entity's state, as opposed to first selecting the current state and only updating if needed, which affects how `Optimistic Locking` occurs. For this reason maybe we should leave the option of using the `Session#update` method to users who want to have this type of behavior.

As an additional note, using `Session#update` with `SelectBeforeUpdate` mapping - which is also deprecated - the `Version` increase does not happen because, like when using `merge`, the `Version` of the parent is not increased since there was no change to its state. Also, `SelectBeforeUpdate` is needed for `orphanRemoval` to work correctly with `update`, while merge handles it correctly.

**Comment by gavinking** [comment]
Yeah `merge()` has very different semantics and so there’s no way to precisely reproduce the behavior of the old `update()` method.

But we didn’t deprecate it because it’s redundant. We deprecated it because it’s bug-prone and something we simply don’t want people to use anymore in new code. (On the other hand, we’re not actually *removing it* yet.)

**Reply by beikov** [reply]
Ok, linking the [original discussion](https://github.com/hibernate/hibernate-orm/discussions/4520) for reference, I think we should at least add a some detail to the deprecation note. `StatelessSession#update` might have the desired semantics if no orphan removal or reattaching (and hence no select before update) is wanted.

**Reply by gavinking** [reply]
Well the other possibility is to use `LockMode.PESSIMISTIC_FORCE_INCREMENT` to force a version update.

**Comment by yuan2006** [comment]
```
    @Transactional(rollbackFor = Exception.class)
    public <T> void batchUpdate(List<T> entities) {
        Session session = this.entityManager.unwrap(Session.class);
        session.setJdbcBatchSize(batchSize);
        for (T t : entities) {
            session.update(t);
        }
    }
```
When I saw the `Session#update` is deprecated, I updated my code to `Session#merge`. To update 1000 records to database by this code, it costs only 0.4 seconds. But it costs 10 seconds when `session.update(t)` is replaced by `session.merge(t)`. `Session#update` performs very bad. So I reverted by code to  `session.update(t)`. 
I'd just to say, don't deprecate `Session#update` and don't remove it in the future version.

**Reply by gavinking** [reply]
Use `StatelessSession.update()`. Stateful sessions are inappropriate for this kind of thing.

**Reply by yuan2006** [reply]
But I can't get a StatelessSession bean by EntityManager bean. How can I get a StatelessSession bean in the Spring application context? The Version is Spring Boot 3.2.3, and the Hibernate is 6.4.4.Final. Thanks.

**Reply by gavinking** [reply]
Well you don't have an `update()` operation on `EntityManager` either so there's no difference there.

I'm afraid I can't help you with your Spring problems, since I'm most definitely not a Spring user.

**Reply by gavinking** [reply]
> Well you don't have an `update()` operation on `EntityManager` either so there's no difference there.

Oh, well, fine, you're using `unwrap()` for that.

So then, `unwrap()` the `SessionFactory` from the `EntityManagerFactory` and create a `StatelessSession`. Easy.

**Reply by yuan2006** [reply]
session#update
```
    public <T> void batchUpdate(List<T> entities) {
        check(entities);
        Session session = this.entityManager.unwrap(Session.class);
        session.setJdbcBatchSize(batchSize);
        for (T t : entities) {
            session.update(t);
        }
    }
```
 batch update 1000 records, costs 0.496 second

but,
statelessSession#update
```
    public <T> void statelessBatchUpdate(List<T> entities) {
        check(entities);
        SessionFactory sessionFactory = this.entityManager.getEntityManagerFactory().unwrap(SessionFactory.class);
        StatelessSession statelessSession = sessionFactory.openStatelessSession();
        statelessSession.setJdbcBatchSize(batchSize);
        for (T t : entities) {
            statelessSession.update(t);
        }
    }
```
batch update 1000 records , costs 8.67 seconds.

Could you review the code, and optimize it?

**Reply by gavinking** [reply]
> Could you review the code, and optimize it?

Sure, of course, my rate is €300 per hour, and it might take me a couple of days work, but I'm at your service.

Alternatively, you could be a good citizen of this open source community, and contribute a solution yourself. Might work out cheaper for you.

**Comment by TatyOko28** [comment]
## Cause Analysis

### JPA Specification (Section 3.4.2 – Version Attributes)

> The version attribute is updated by the persistence provider runtime when the object is written to the database. **All non-relationship fields and properties and all relationships owned by the entity are included in version checks.**

### JPA Specification (Section 2.10 – Entity Relationships)

> The inverse side of a bidirectional relationship must refer to its owning side by use of the `mappedBy` element of the `OneToOne`, `OneToMany`, or `ManyToMany` annotation. The `mappedBy` element designates the property or field in the entity that is the owner of the relationship.

### What This Means

| Side | Version Check | Example |
|------|---------------|---------|
| **Owning side** (holds foreign key) |  Included in version check | `@ManyToOne` side |
| **Inverse side** (has `mappedBy`) |  NOT included in version check | `@OneToMany(mappedBy=...)` |

**In the user's case:**
- `Parent` is the **inverse side** (`@OneToMany(mappedBy = "parent")`)
- Changing the child collection does **not** affect the parent's version according to JPA

### When Ownership Changes

```java
// Remove mappedBy – Parent becomes owning side
@OneToMany  // No mappedBy
@JoinColumn(name = "parent_id")  // Requires join column
private List<Child> children = new ArrayList<>();
```

With this change, `merge()` **does** increment the version when children change.

## Behavior Comparison: update() vs merge()

### Scenario: Change only child collection (inverse side)

| Method | Hibernate 5 | Hibernate 6 | Version Increment? |
|--------|-------------|-------------|-------------------|
| `session.update()` |  Version increments |  Deprecated | **Yes** (forces UPDATE) |
| `session.merge()` | N/A (same as 6) |  Version NOT incremented | **No** (per JPA spec) |

### Scenario: Change parent entity field

| Method | Version Increment? |
|--------|-------------------|
| `session.update()` |  Yes |
| `session.merge()` |  Yes |

### Scenario: SelectBeforeUpdate mapping enabled

| Method | With SelectBeforeUpdate | Version Increment? |
|--------|------------------------|-------------------|
| `session.update()` |  Enabled | No (no parent changes detected) |
| `session.merge()` | N/A | No (always checks first) |
